### PR TITLE
Expand "Local development"

### PR DESCRIPTION
**Pull request by tbeseda**
> 🚧   **WIP**: creating draft PR to coordinate with Ryan, but all feedback welcome :)

### Expand "_Local development_"

I added an example for "_Creating a new resource_". I used a hypothetical "hit counter" example. `arc init` is used for scaffolding (mentioned in [this issue](https://github.com/architect/arc.codes/issues/312))
This type of workflow was a point of interest as I was building my first app.

Stub out "_Debugging_" section that may or may not be necessary. Leads into testing.

### Move "_Testing_" to its own document

Testing is definitely a part of "local dev" but probably warrants its own doc.

## More ideas for "Local development"

* provide a short implementation of the "hit counter" event feature, just to tie it all together
* use inline tabs to demonstrate Ruby and Python alongside the Node example
* elaborate on what Architect Sandbox emulates locally (and how?)
  * ie. queues, events, dynalite
  * this is covered in `reference/cli/sandbox`, but there might be an opportunity to tout some of the capabilities without requiring the reader to dig too far.

~I avoided second person, but after [research](https://docs.microsoft.com/en-us/style-guide/grammar/person) on [style guides](https://developers.google.com/style/person) by [large orgs](https://developer.salesforce.com/docs/atlas.en-us.salesforce_pubs_style_guide.meta/salesforce_pubs_style_guide/style_second_person.htm), I'll change it.~

---

- [x] Forked the repo and created your branch from `master`
- [x] Made sure tests pass (run `npm it` from the repo root)
- [x] Updated relevant documentation internal to this repo (e.g. `readme.md`, help docs, inline docs & comments, etc.)
- [x] Linked to any related issues, PRs, etc. below that may relate to, consume, or necessitate these changes

### CODE REVIEW

File: src/http/any-catchall/index.js [hunk 1]
```diff
@@ -1,14 +1,15 @@
 let { http } = require('@architect/functions')
+let asap = require('@architect/asap')
 
 // middleware to preserve old urls
 let redirect = require('./redirect')
 
 // middleware proxy s3 assets
-let asap = http.proxy({
+let static = asap({
```
Comment (ryanblock) [inline src/http/any-catchall/index.js:8]: Fyi, careful with `static`, it's reserved! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#future_reserved_keywords

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 2]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
+<!-- unsure if this diagram is necessary -->
+```
+.
+├── src
+│   └── events
+│       └── hit-counter
+│           ├── index.js
+│           └── config.arc
+│
+└── app.arc
```
Comment (ryanblock) [inline src/views/docs/en/guides/developer-experience/local-development.md:67]: Hmm, some funky unicode artifacts here?
image

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 3]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
-  t.ok(result, 'got 200 response')
-})
+Your new event may require some dependencies. For this example, [@architect/functions](https://github.com/architect/functions) will be helpful to handle the event subscription. Dependencies can be installed directly to the `hit-counter` directory:
 
-test('teardown', async t=> {
-  t.plan(1)
-  await sandbox.end()
-  t.ok(true, 'sandbox ended')
-})
+1. Inside `./src/events/hit-counter` create a `package.json` file with an empty `dependencies` entry:
```
Comment (ryanblock) [inline src/views/docs/en/guides/developer-experience/local-development.md:74]: Good news, we do Lambda treeshaking now! No need to have individual `package.json` files, just add `@architect/functions` to root `package.json` and we'll pick it up at hydration / deploy time.

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 4]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
-  await sandbox.end()
-  t.ok(true, 'closed')
-})
+3. From the project root start up the sandbox if it is not already running:
+
+```console
+arc sandbox
 ```
+
+Architect will hydrate your [shared code](/docs/en/guides/developer-experience/sharing-code) and run the updated Architect project.
```
Comment (ryanblock) [inline src/views/docs/en/guides/developer-experience/local-development.md:96]: We didn't mention adding shared code as part of this local dev guide, so this is a bit awkward. But it'd def make sense to demonstrate shared code.

File: test/sandbox-http-test.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+let tiny = require('tiny-json-http')
+let sandbox = require('@architect/sandbox')
+
+const host = 'http://localhost:3333'
+
+test('sandbox HTTP request', async t => {
+  t.plan(4)
+
+  await sandbox.start({ quiet: true })
+  t.ok(true, `sandbox started on ${host}`)
```
Comment (ryanblock) [inline test/sandbox-http-test.js:11]: `t.ok(true)` → `t.pass()`

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 6]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
-  t.ok(true, 'sandbox started on http://localhost:3333')
-})
+### Add dependencies (optional)
 
-test('get /', async t=> {
-  t.plan(1)
-  let result = await tiny.get({ url: 'http://localhost:3333' })
-  t.ok(result, 'got 200 response')
-})
+Your new event may require some dependencies. For this example, [@architect/functions](https://github.com/architect/functions) will be helpful to handle the event subscription. Dependencies can be installed directly to the `hit-counter` directory:
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/local-development.md:72]: Maybe we can link to the [dependency management guide](https://arc.codes/docs/en/guides/developer-experience/dependency-management) in here somehow to note the two approaches to dependency management? 
Perhaps something like:

```suggestion
Your new event may require some dependencies. For this example, [the @architect/functions runtime helper library](../../reference/runtime/node.js) will be helpful to handle the event subscription. [Dependencies in node.js can be managed](dependency-management) either globally from the root `package.json`, or managed via separate `package.json` files in each function's directory. In this example, let's manage them directly in the event's `src/events/hit-counter` directory:
```

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 7]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
-  let cats = await data.cats.scan({})
-  t.ok(Array.isArray(cats.Items), 'bag o cats')
-})
-
-test('teardown', async t=>{
-  t.plan(1)
-  await sandbox.end()
-  t.ok(true, 'closed')
-})
+3. From the project root start up the sandbox if it is not already running:
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/local-development.md:90]:
```suggestion
3. From the project root start up the [sandbox](../../reference/cli/sandbox) if it is not already running:
```

File: src/views/docs/en/guides/developer-experience/local-development.md (hunk 4 above)
Comment (filmaj) [reply src/views/docs/en/guides/developer-experience/local-development.md:96]: Agreed, perhaps we can reword this to state that _sandbox_ will hydrate function code and link to the `arc hydrate` docs while we are at it?

```suggestion
[`arc sandbox`](../../reference/cli/sandbox) will automatically [hydrate - that is, install needed dependencies](../../reference/cli/hydrate) - for every function making up your Architect project.
```

File: src/views/docs/en/guides/developer-experience/local-development.md [hunk 8]
```diff
@@ -4,112 +4,103 @@ category: Developer experience
...
+
+> 📖 Read more about [dependency management](/docs/en/guides/developer-experience/dependency-management).
+
+## Debugging
+
+With an Architect project running locally, there are several ways to work on your application.
+
+You can run a front-end application from the same Architect project to communicate with the back-end or use a client to interface with HTTP functions.
+
+Of course, the best way to catch bugs is by [testing your Architect project](/docs/en/guides/developer-experience/testing).
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/local-development.md:106]:
```suggestion
Of course, the best way to catch bugs is by [testing your Architect project](testing).
```

File: src/views/docs/en/guides/developer-experience/testing.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
+---
+title: Testing
+category: Developer experience
+description: How to test with Architect
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/testing.md:4]:
```suggestion
description: How to test your Architect project
```

File: src/views/docs/en/guides/developer-experience/testing.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+---
+
+### `@http` functions
+
+Use `@architect/sandbox` as a module to test `@http` functions with Node:
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/testing.md:9]:
```suggestion
Use `@architect/sandbox` as a module to test [`@http`](../../reference/app.arc/http) functions with Node:
```

File: src/views/docs/en/guides/developer-experience/testing.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+test('teardown', async t=> {
+  t.plan(1)
+  await sandbox.end()
+  t.ok(true, 'sandbox ended')
+})
+```
+
+### `@events` and `@queues`
+
+Use `@architect/functions` to publish locally to `@events` or `@queues` for testing.
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/testing.md:37]:
```suggestion
Use [`@architect/functions`](../../reference/runtime/node.js) to publish locally to [`@events`](../../reference/app.arc/events) or [`@queues`](../../reference/app.arc/queues) for testing.
```

File: src/views/docs/en/guides/developer-experience/testing.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  t.plan(1)
+  await sandbox.end()
+  t.ok(true, 'closed')
+})
+
+```
+
+### `@tables`
+
+Using `@architect/functions` to interact with DynamoDB `@tables` defined locally.
```
Comment (filmaj) [inline src/views/docs/en/guides/developer-experience/testing.md:75]:
```suggestion
Using `@architect/functions` to interact with DynamoDB [`@tables`](../../reference/app.arc/tables) defined locally.
```

**Review by filmaj — CHANGES_REQUESTED** [review]
Wow, this is so awesome! Thanks for taking this on! I think this is a massive improvement!

I left a few comments, mostly around wording and dropped a few suggestions around linking to other parts of the docs from relevant areas.

Thank you so much for sending this PR!

**Comment by tbeseda** [comment]
Awesome feedback @ryanblock and @filmaj 
I'm going to synthesize this today and update here this afternoon.
Thanks for taking a look.

File: src/views/docs/en/guides/developer-experience/local-development.md (hunk 2 above)
Comment (tbeseda) [reply src/views/docs/en/guides/developer-experience/local-development.md:67]: Super weird. that line is the same as the other `└── app.arc` instances in project-layout.md and queues.md.  
and the `─` character that shows up garbled in your screenshot is used in each of the lines above the app.arc line in your screenshot.

The issue must be between the code and the HTML output, maybe the Markdown compiler. but I'm having a hard time reproducing that garbled text.  
I'm already taking a look there since there seems to be an issue with my highlighter.js change, so I can take some time to look more at this.

**Comment by tbeseda** [comment]
I included suggested edits and decided to use the project's main `package.json` when demonstrating dependency management.

I also added `markdown-it-attrs` in order to add `target="_blank"` to links that leave arc.codes. This will also be useful for adding other attributes while continuing to stick with Markdown instead of too much HTML.  
also added a bit of CSS to show an icon next to external links. I can update all external links in a future PR if maintainers like this idea.

external link

Last, Ryan mentioned an issue with the encoding on characters used in the folder structure diagram (see his screenshot above). I wasn't able to reproduce this. I tried with Node 12, 14, and 16. Also fresh installs of the repo after clearing my npm cache.  
It may relate to the occasional refresh issue I've seen when the sandbox restarts and the browser uses locally cached assets. Maybe another contributor can pull and boot up the app to check.

**Comment by tbeseda** [comment]
I'm closing this as we're making more targeted passes on docs. I will maintain the branch as a source for improvements to the text of the doc. Some changes here add dependencies and might be out of scope in the near future.
